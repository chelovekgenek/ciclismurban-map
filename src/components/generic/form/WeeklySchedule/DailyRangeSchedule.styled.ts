import styled from "styled-components"
import { Button } from "components/generic/ui"

export const EmptyBox = styled.div`
  text-align: center;
`

export const ActionWrapper = styled(Button.Group)`
  float: right;
`
