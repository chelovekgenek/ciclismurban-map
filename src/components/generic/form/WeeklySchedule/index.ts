export { WeeklyScheduleWrapper as WeeklySchedule } from "./WeeklySchedule"
