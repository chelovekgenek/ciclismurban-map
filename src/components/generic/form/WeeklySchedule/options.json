{
  "format": "HH:mm",
  "titles": {
    "mon": "Пн",
    "tue": "Вт",
    "wed": "Ср",
    "thu": "Чт",
    "fri": "Пт",
    "sat": "Сб",
    "sun": "Вс"
  }
}
