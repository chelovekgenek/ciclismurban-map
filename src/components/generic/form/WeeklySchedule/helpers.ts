import moment, { Moment } from "moment"

import options from "./options.json"

export const parse = (v: string) => moment(v, options.format)
export const format = (v: Moment) => v.format(options.format)
