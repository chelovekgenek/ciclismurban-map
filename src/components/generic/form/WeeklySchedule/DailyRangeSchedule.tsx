import React, { useMemo, useCallback } from "react"
import { TimePicker } from "antd"
import { Moment } from "moment"

import { Grid } from "components/generic/layout"
import { Button } from "components/generic/ui"
import { WeeklyScheduleModel } from "models/location"

import * as Styles from "./DailyRangeSchedule.styled"
import options from "./options.json"
import { parse } from "./helpers"

interface IProps {
  value?: WeeklyScheduleModel[keyof WeeklyScheduleModel][number]
  onAdd?: () => void
  onCancel?: () => void
  onChange?: (value: Required<IProps>["value"]) => void
}

export const DailyRangeSchedule: React.FC<IProps> = ({ value, onAdd, onCancel, onChange }) => {
  const from = useMemo(() => (value ? parse(value.from) : undefined), [value])
  const to = useMemo(() => (value ? parse(value.to) : undefined), [value])
  const handleFromChange = useCallback(
    (_time: Moment, timeStr: string) =>
      onChange &&
      onChange({
        from: timeStr,
        to: value!.to,
      }),
    [value, onChange],
  )
  const handleToChange = useCallback(
    (_time: Moment, timeStr: string) =>
      onChange &&
      onChange({
        from: value!.from,
        to: timeStr,
      }),
    [value, onChange],
  )
  return (
    <Grid.Row>
      <Grid.Col span={9}>
        {from ? (
          <TimePicker format={options.format} value={from} onChange={handleFromChange} />
        ) : (
          <Styles.EmptyBox>-</Styles.EmptyBox>
        )}
      </Grid.Col>
      <Grid.Col span={9}>
        {to ? (
          <TimePicker format={options.format} value={to} onChange={handleToChange} />
        ) : (
          <Styles.EmptyBox>-</Styles.EmptyBox>
        )}
      </Grid.Col>
      <Grid.Col span={6}>
        <Styles.ActionWrapper>
          <Button icon="minus" onClick={onCancel} disabled={!onCancel} />
          <Button icon="plus" onClick={onAdd} disabled={!onAdd} />
        </Styles.ActionWrapper>
      </Grid.Col>
    </Grid.Row>
  )
}
