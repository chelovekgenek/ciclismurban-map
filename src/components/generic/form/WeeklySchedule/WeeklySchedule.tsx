import React, { useMemo } from "react"
import { Field, FieldProps } from "formik"
import { pullAt, cloneDeep } from "lodash-es"
import { WeeklyScheduleModel } from "@ciclismurban/models"

import { Grid } from "components/generic/layout"
import { FormItem } from "../FormItem"

import options from "./options.json"
import { DailyRangeSchedule } from "./DailyRangeSchedule"

const defaultValue = {
  from: "08:00",
  to: "20:00",
}

interface IProps {
  name: string
  label?: string
}

type TDailyRangeSchedule = Required<React.ComponentProps<typeof DailyRangeSchedule>>

export const WeeklyScheduleWrapper: React.FC<IProps> = ({ name, label }) => (
  <Field name={name}>{(fieldProps: FieldProps) => <WeeklySchedule {...fieldProps} label={label} name={name} />}</Field>
)

export const WeeklySchedule: React.FC<FieldProps & IProps> = ({ field, form, label }) => {
  const list = useMemo(
    () =>
      Object.entries(options.titles).map(([key, title]) => {
        const values = field.value[key] as WeeklyScheduleModel[keyof typeof options.titles]
        const setFieldValue = (nextValues: typeof values) => {
          form.setFieldValue(field.name, { ...field.value, [key]: nextValues })
          form.setFieldTouched(field.name, true)
        }
        const handleAdd = () => {
          const nextValues = values.concat(defaultValue)
          setFieldValue(nextValues)
        }
        const handleCancel = (index: number) => () => {
          const nextValues = cloneDeep(values)
          pullAt(nextValues, [index])
          setFieldValue(nextValues)
        }
        const handleFieldChange = (index: number): TDailyRangeSchedule["onChange"] => v => {
          const nextValues = cloneDeep(values)
          nextValues.splice(index, 1, v)
          setFieldValue(nextValues)
        }
        return (
          <Grid.Row key={key}>
            <Grid.Col span={3}>{title}</Grid.Col>
            <Grid.Col span={21}>
              {values.length ? (
                Object.values(values).map((item, i) => (
                  <DailyRangeSchedule
                    key={i}
                    value={item}
                    onChange={handleFieldChange(i)}
                    onCancel={handleCancel(i)}
                    {...{
                      onAdd: i === values.length - 1 ? handleAdd : undefined,
                    }}
                  />
                ))
              ) : (
                <DailyRangeSchedule onAdd={handleAdd} />
              )}
            </Grid.Col>
          </Grid.Row>
        )
      }),
    [field.value, field.name, form],
  )
  return (
    <FormItem label={label} touched={form.touched[field.name]} error={form.errors[field.name]}>
      {list}
    </FormItem>
  )
}
