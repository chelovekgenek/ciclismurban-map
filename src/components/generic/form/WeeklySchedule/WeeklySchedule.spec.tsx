import React from "react"
import { shallow } from "enzyme"

import fieldProps from "mocks/formik-field-props"
import weeklySchedule from "mocks/weekly-schedule.json"

import { WeeklySchedule } from "./WeeklySchedule"

describe("pages/WeeklySchedule", () => {
  const render = (props: Partial<React.ComponentProps<typeof WeeklySchedule>> = {}) =>
    shallow(
      <WeeklySchedule
        {...fieldProps}
        field={{ ...fieldProps.field, value: weeklySchedule, name: "weekly-schedule" }}
        name="weekly-schedule"
        label="weekly-schedule"
        {...props}
      />,
    )
  it("should match snapshot", () => {
    expect(render()).toMatchSnapshot()
  })
})
