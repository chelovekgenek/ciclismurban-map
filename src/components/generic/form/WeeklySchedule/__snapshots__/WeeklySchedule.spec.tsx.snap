// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pages/WeeklySchedule should match snapshot 1`] = `
<FormItem
  label="weekly-schedule"
>
  <Row
    gutter={0}
    key="mon"
  >
    <Col
      span={3}
    >
      Пн
    </Col>
    <Col
      span={21}
    >
      <DailyRangeSchedule
        key="0"
        onAdd={[Function]}
        onCancel={[Function]}
        onChange={[Function]}
        value={
          Object {
            "from": "08:00",
            "to": "20:00",
          }
        }
      />
    </Col>
  </Row>
  <Row
    gutter={0}
    key="tue"
  >
    <Col
      span={3}
    >
      Вт
    </Col>
    <Col
      span={21}
    >
      <DailyRangeSchedule
        key="0"
        onAdd={[Function]}
        onCancel={[Function]}
        onChange={[Function]}
        value={
          Object {
            "from": "08:00",
            "to": "20:00",
          }
        }
      />
    </Col>
  </Row>
  <Row
    gutter={0}
    key="wed"
  >
    <Col
      span={3}
    >
      Ср
    </Col>
    <Col
      span={21}
    >
      <DailyRangeSchedule
        key="0"
        onAdd={[Function]}
        onCancel={[Function]}
        onChange={[Function]}
        value={
          Object {
            "from": "08:00",
            "to": "20:00",
          }
        }
      />
    </Col>
  </Row>
  <Row
    gutter={0}
    key="thu"
  >
    <Col
      span={3}
    >
      Чт
    </Col>
    <Col
      span={21}
    >
      <DailyRangeSchedule
        key="0"
        onAdd={[Function]}
        onCancel={[Function]}
        onChange={[Function]}
        value={
          Object {
            "from": "08:00",
            "to": "20:00",
          }
        }
      />
    </Col>
  </Row>
  <Row
    gutter={0}
    key="fri"
  >
    <Col
      span={3}
    >
      Пт
    </Col>
    <Col
      span={21}
    >
      <DailyRangeSchedule
        onAdd={[Function]}
      />
    </Col>
  </Row>
  <Row
    gutter={0}
    key="sat"
  >
    <Col
      span={3}
    >
      Сб
    </Col>
    <Col
      span={21}
    >
      <DailyRangeSchedule
        key="0"
        onCancel={[Function]}
        onChange={[Function]}
        value={
          Object {
            "from": "08:00",
            "to": "12:00",
          }
        }
      />
      <DailyRangeSchedule
        key="1"
        onAdd={[Function]}
        onCancel={[Function]}
        onChange={[Function]}
        value={
          Object {
            "from": "14:00",
            "to": "20:00",
          }
        }
      />
    </Col>
  </Row>
  <Row
    gutter={0}
    key="sun"
  >
    <Col
      span={3}
    >
      Вс
    </Col>
    <Col
      span={21}
    >
      <DailyRangeSchedule
        onAdd={[Function]}
      />
    </Col>
  </Row>
</FormItem>
`;
